cranelift-codegen/meta/src/shared/legalize.rs

```diff
@@ -272,6 +273,25 @@ pub(crate) fn define(insts: &InstructionGroup, imm: &Immediates) -> TransformGro
         ],
     );
 
+    // TODO(ryzokuken): explore the perf diff w/ x86_umulx and consider have a
+    // separate legalization for x86.
+    for &(ty, ty_half) in &[(I64, I32), (I128, I64)] {
+        narrow.legalize(
+            def!(a = imul.ty(x, y)),
+            vec![
+                def!((xl, xh) = isplit.ty(x)),
+                def!((yl, yh) = isplit.ty(y)),
+                def!(al = imul.ty_half(xl, yl)),
+                def!(a1 = imul.ty_half(xh, yl)),
+                def!(a2 = imul.ty_half(xl, yh)),
+                def!(a3 = umulhi.ty_half(xl, yl)),
+                def!(a4 = iadd.ty_half(a1, a2)),
+                def!(ah = iadd.ty_half(a4, a3)),
+                def!(a = iconcat.ty_half(al, ah)),
```

Comment:
The type params can be left out as they are inferrable, right?

Reply:
Agreed, and it should make the legalization easier to read too.

Reply:
Can you also swap the order of definitions of a3 and a4, to reduce the corresponding SSA values' lifetimes, please?

Reply:
@bnjbvr on it.